Proxy, Compression, More Functionality

MrTechno opened the pull request:
Hi oxan,

I've cut the commits into three sets; one for GZIP support, one to correct the web stream results from behind an incoming proxy and another for EditSchedule. The last one is very much a work in progress since it only has the schedule type parameter even though there's plenty of other things you could alter. Take a look when you've got some time and let me know what you think.
- Mike

MrTechno committed:
Added gzip support

MPExtended will compress the response if the Accept-Encoding header includes gzip

MrTechno committed:
Added X-Original-URL check

When MPExtended is behind a proxy it needs to know what the original URL was so web stream results can be set correctly

MrTechno committed:
Added EditSchedule

WebMediaPortal uses two operations (DeleteSchedule + AddSchedule) to simulate an edit. This makes it possible to do the same thing in one operation which is important if you're making the request remotely.

MrTechno committed:
Fixed X-Original-URL

The default port logic was missing brackets around the HTTP or HTTPS
condition

MrTechno committed:
Changed X-Original-URL to X-Original-Host

It turns out URLRewrite in IIS already adds an X-Original-URL header to the proxy request but it contains only the server path which is exactly the opposite bit that I want. That being the case I've changed the header MPE looks for to X-Original-Host

MrTechno committed:
Fixed "System.InvalidOperationException: Collection was modified" during service startup

As noted by Ollienator in
https://github.com/MPExtended/MPExtended/commit/25f93d263dde1cf37b80344a791eafb63fb158c7#diff-a91b3c24b081341356d4fe00121e62cc
the following error is sometimes thrown:

System.InvalidOperationException: Collection was modified; enumeration
operation may not execute.
at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource
resource)
at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
at System.Collections.Generic.List`1.Enumerator.MoveNext()
at System.Linq.Enumerable.WhereListIterator`1.MoveNext()
at System.Linq.Enumerable.<SelectManyIterator>d__14`2.MoveNext()
at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
at System.Linq.OrderedEnumerable`1.<GetEnumerator>d__0.MoveNext()
at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source)
at
MPExtended.Libraries.Service.Network.NetworkInformation.LoadIPAddress()
in
E:\Source\MPExtended\Libraries\MPExtended.Libraries.Service\Network\NetworkInformation.cs:line
98
at
MPExtended.Libraries.Service.Network.NetworkInformation.GetIPAddressForUri()
in
E:\Source\MPExtended\Libraries\MPExtended.Libraries.Service\Network\NetworkInformation.cs:line
78
at
MPExtended.Libraries.Service.Hosting.BaseAddresses.GetForService(String
serviceName) in
E:\Source\MPExtended\Libraries\MPExtended.Libraries.Service\Hosting\BaseAddresses.cs:line
36
at MPExtended.Libraries.Service.Hosting.WCFHost.CreateHost(Plugin`1
plugin) in
E:\Source\MPExtended\Libraries\MPExtended.Libraries.Service\Hosting\WCFHost.cs:line
86

Since the order of the "Loading service " logging output is not always
the same I suspect the services are loading in parallel causing a race
condition where NetworkInformation::GetNetworkInterfaces initialises
_interfaces more than once. To fix this I've put a lock guard around the
contents of the method so subsequent threads will block until
_interfaces is fully set up

oxan commented on Libraries/MPExtended.Libraries.Service/Compression/GZipMessageEncoderFactory.cs:
```diff
@@ -0,0 +1,249 @@
+        {
+            //Log.Debug(GetType().FullName + "::Detach " + owner.GetType().FullName);
+        }
+    }
```
Since this class is empty (and probably unused now), get rid of it.

MrTechno replied:
GZip should only be applied if it's listed in the Accept-Encoding request header. GZioMessageInspector performs this check and GzipContext is the way of telling GZipMessageEncoder to apply compression.

MrTechno replied:
There's other potential improvements here such as only compressing the response content if the length is over a certain size. I haven't worked out how to do it yet though.

oxan replied:
Makes sense. Would be nice if you could add a comment with that.

oxan commented:
Hi,

Sorry for the late reply. I try to answer Pull Requests fast, as it's the least you can do for people contributing to your software, but that obviously failed. I'm a bit too busy for MPExtended nowadays. 

The idea behind and the general architecture of your changes are quite good. I've a few minor comments on the code itself though, if you could fix those I'll merge it! Thanks in advance.

oxan commented on Libraries/MPExtended.Libraries.Service/Compression/GZipMessageEncoderFactory.cs:
```diff
@@ -0,0 +1,249 @@
+        {
+            if (messageEncoderFactory == null)
+                throw new ArgumentNullException("messageEncoderFactory", "A valid message encoder factory must be passed to the GZipEncoder");
+            //Log.Debug(GetType().FullName + "::ctor with " + messageEncoderFactory.GetType().FullName);
```
It'd be nice if you could cleanup commented-out statements and properties etc.

MrTechno replied:
Done

oxan commented on Libraries/MPExtended.Libraries.Service/Compression/GZipMessageEncodingBindingElement.cs:
```diff
@@ -0,0 +1,117 @@
+            else 
+            {
+                return base.GetProperty<T>(context);
+            }
```
What's the reason requests for XmlDictionaryReaderQuotas are forwarded to the inner binding, and all others to the parent?

MrTechno replied:
You know what? I don't actually know. Some long forgotten debugging attempt of mine perhaps. I'll see if it works without it.

MrTechno replied:
Seems to work without the dictionary check.Removed entire GetProperty method.

oxan commented on Libraries/MPExtended.Libraries.Service/Compression/GZipMessageEncoderFactory.cs:
```diff
@@ -0,0 +1,249 @@
+using System.ServiceModel.Description;
+using MPExtended.Libraries.Service;
+
+namespace Microsoft.ServiceModel.Samples
```
Is there any reason why this class isn't in the MPExtended.Libraries.Service.Compression namespace? You could even put them in the MPExtended.Libraries.Service.WCF namespace, where our other WCF extensions are.

MrTechno replied:
Just an overhang from the place I got the code. I'll move the binding element to WCF and change the factory namespace

MrTechno replied:
Done

oxan commented on Libraries/MPExtended.Libraries.Service/Network/NetworkInformation.cs:
```diff
@@ -38,34 +38,38 @@ public struct NetworkInterfaceData
-                    Addresses = new List<IPAddress>()
-                };
+                if (_interfaces != null)
+                    return _interfaces;
```
You could keep this check outside the lock, it'll speed up concurrent method calls and locking is only needed when you change the list anyway.

MrTechno replied:
The GitHub diff doesn't show the change very clearly. I reason I put a lock around everything because _interfaces is instantiated then populated so checking for not null is no guarantee of avoiding the concurrent modification problem.

oxan replied:
Hmm, that's true. Nevermind then.

oxan commented on Libraries/MPExtended.Libraries.Service/WCF/GZipBehavior.cs:
```diff
@@ -0,0 +1,87 @@
+                var accept = prop.Headers[HttpRequestHeader.AcceptEncoding];
+                if (!string.IsNullOrEmpty(accept) && accept.Contains("gzip"))
+                {
+                    Log.Debug(GetType().FullName + "::AfterReceiveRequest enable gzip for this request");
```
Would be nice to change this to a Trace-level log; we don't want a log message to end up in the log file for each service call performed on users' systems (same in BeforeSendReply below).

MrTechno replied:
Done

oxan commented on Libraries/MPExtended.Libraries.Service/WCFUtil.cs:
```diff
@@ -51,6 +52,36 @@ public static string GetCurrentRoot()
+                            {
+                                portStr = ":" + uri.Port;
+                            }
+                            return String.Format("{0}://{1}{2}/MPExtended/", uri.Scheme, uri.Host, portStr);
```
Not sure how Uri behaves, does this still work with `X-Original-Host: some.host`? Or do you need to specify the scheme in the header? In the second case, i'd choose a different name for the header (something like X-Original-URL).

MrTechno replied:
You need to know the original scheme to create a self referencing URL and the Host header does not supply it. X-Original-URL is already something that URLRewrite uses which is why I changed it. It doesn't actually matter what the header is called as long as it's consistent between what WCFUtil is looking for and how the incoming server is configured. Maybe even something app specific like X-MPExtended-URL?

oxan replied:
Makes sense. I'd name it X-Root-URL then, but X-MPExtended-URL is fine too.

MrTechno replied:
Now changed to X-Root-URL. When there's an official MPE release which includes this I'll add HOWTOs for IIS and Apache to the wiki pages.

oxan commented on Services/MPExtended.Services.TVAccessService.Interfaces/ITVAccessService.cs:
```diff
@@ -108,6 +108,10 @@ public interface ITVAccessService
 
         [OperationContract]
         [WebGet(ResponseFormat = WebMessageFormat.Json)]
+        WebBoolResult EditSchedule(int scheduleId, WebScheduleType scheduleType);
```
Could you add all the arguments of AddSchedule here? You don't have to implement them, but extending the argument list later on breaks binary compatability and requires all clients to be recompiled, so I want to have the full argument list here in all stable releases.

MrTechno replied:
I've suddenly realised why both AddSchedule and AddScheduleDetailed exist :-)

I will do EditSchedule and EditScheduleDetailed in the same way, unless you can think of a way to leave any fields that are not supplied as the existing values?

oxan replied:
You could make all parameters nullable and only change the non-null fields. Not sure whether we want to be able to set a field to null though.

MrTechno replied:
Done

MrTechno committed:
Refactored class names from GZIP to Compression

The only GZIP specific code is in the message encoder, I've added
deflate support as well so now the classes have misleading names,
CompressionContext (formerly GzipContext) now carries compression type
information so it actually has something to do in addition to just
existing. It also solves the awkward problem of not knowing if Gzip or
GZip is the correct capitalisation.

MrTechno committed:
Changed X-Original-Host to X-Root-URL

Put "URL" back in to imply it's more than just the hostname (which you
can get as standard anyway), changed log level to trace as the end user
is unlikely to be interested every time once it's been set up.

MrTechno committed:
Compression fixes

For some reason the streams only flush if you put them inside "using"
blocks. I've tried not to do too much cargo cult copy 'n' paste, will
revisits this code later

MrTechno committed:
Enhanced EditSchedule

Included all parameters you can pass to AddScheduleDetailed. It's a mix
of default null values optional args where the language syntax allows
(this is where my lack of C# experience shows I'm afraid)

MrTechno commented:
I've added deflate support to the compression namespace but I'm having real trouble finding a program that will decompress it. May end up writing one just to make sure it works. I've removed the commented out logging and what remains has been changed to trace level and reworked to use the parameterised style instead of concatenating everything together as a string.

MrTechno committed:
Improved LibraryList error logging

Added the type so "unknown id 0" becomes "unknown id 0 of type Music" in
the logs

MrTechno committed:
Changed EditSchedule to Nullable ints

The int parameters were defaulted to int.MinValue, now changed to
Nullable objects. The strings stay as they were since the defaults are
already null and that remains the check in the logic.

MrTechno committed:
Fixed compression for downloads and streaming

More of the endpoints in App.config needed to be wired up to use
compressionBehavior and customWebHttpBinding. Tidied up
CompressionMessageEncoderFactory, removed commented out code and
DecompressBuffer which is never called

MrTechno committed:
Added content type to CompressionMessageInspector logging in BeforeSendReply
